README.md
LFT use ICMP time exceeded packets (TTL expired in transit) to measuring latency. PING use ICMP Echo Request packets to measure latency. LFTPING can get echo hop latency by PING and LFT methods at the same time. Like as follow,

{"tool":"lft","ip":"112.1.32.1","protocol":"icmp","trace":[
{"type":"icmp no reply from target","note":"Try advanced options to see packets"}
]}
{"tool":"lft","ip":"112.2.32.1","protocol":"icmp","trace":[
{"seq":1,"asn":4837,"net":"APNIC-27/UNICOM-CQ","ip":"27.10.56.1","lfttime":"65.91/30.99/30.93 ms","pingtime":"31.297 ms"}
,{"seq":2,"asn":4837,"net":"APNIC7/UNICOM-CQ","ip":"221.5.246.169","lfttime":"35.76/33.89/31.93 ms","pingtime":"31.399 ms"}
,{"seq":3,"asn":4837,"net":"APNIC7/UNICOM-CQ","ip":"221.5.239.37","lfttime":"70.99/64.12/64.37 ms","pingtime":"67.958 ms"}
,{"seq":4,"asn":4837,"net":"APNIC5/CNCGROUP","ip":"219.158.99.213","lfttime":"58.30/60.32/60.58 ms","pingtime":"60.741 ms"}
,{"seq":5,"asn":4837,"net":"APNIC5/CNCGROUP","ip":"219.158.100.201","lfttime":"97.44/97.23/96.30 ms","pingtime":"98.726 ms"}
,{"seq":6,"asn":4837,"net":"APNIC5/ChinaUnicom-BACKBONE","ip":"219.158.4.78","lfttime":"86.44/85.54/84.53 ms","pingtime":"86.052 ms"}
,{"seq":7,"asn":4837,"net":"APNIC5/CNCGROUP-BACKBONE","ip":"219.158.34.242","lfttime":"84.18/86.13/96.11 ms","pingtime":" ms"}
,{"seq":8,"asn":9808,"net":"APNIC7/CMNET","ip":"221.176.17.33","lfttime":"99.55/121.01 ms"asn":9808,"net":"APNIC7/CMNET","ip":"221.176.17.161","lfttime":"84.92 ms","pingtime":"86.746 ms"}
,{"seq":9,"asn":9808,"net":"APNIC7/CMNET","ip":"221.183.10.125","lfttime":"93.31 ms"asn":9808,"net":"APNIC7/CMNET","ip":"221.183.10.81","lfttime":"93.81 ms"asn":9808,"net":"APNIC7/CMNET","ip":"221.183.10.121","lfttime":"112.74 ms","pingtime":"104.893 ms"}
,{"seq":10,"asn":9808,"net":"APNIC7/CMNET","ip":"221.183.12.10","lfttime":"92.01/96.65 ms","pingtime":" ms"}
,{"seq":11,"asn":56046,"net":"APNIC7/CMNET","ip":"221.181.243.2","lfttime":"112.93 ms","pingtime":"105.957 ms"}
,{"seq":12,"asn":56046,"net":"APNIC4/CMNET-jiangsu","ip":"218.206.97.114","lfttime":"95.83 ms","pingtime":"94.698 ms"}
,{"seq":13,"asn":56046,"net":"APNIC-112/CMNET","ip":"112.2.32.1","lfttime":"90.61 ms","pingtime":"91.809 ms"}
]}


CHANGELOG

LFT CHANGELOG 

lft 3.35 / WhoB 3.5
----------------------
- Roy T. provided some clean-ups to avoid double free()s 
- Bug fixes only


lft 3.33 / WhoB 3.5
----------------------
- Fixed free(sess->hostname) bug (segfault on unresolvable hostname)
- Improved error hanlding of pcap failures related to data link type
- Kurt's FreeBSD fix for pcap snprintf
- Bug fixes only


lft 3.32 / WhoB 3.5
----------------------
- Added support for several encapsulating protocols such as PPP
- no other changes


lft 3.31 / WhoB 3.5
----------------------
- Added #define for AI_NUMERICSERV undeclared on some platforms/versions
- No other changes


lft 3.3 / WhoB 3.5
----------------------
- Improved LFT target detection behind firewalls
- Updated LFT for newer pcap version defaults
- WhoB ignores comments [#,;] in input files
- Added LFT option -f to specify fake source address
- Applied many, many, many fixes from Juli M. of USA -- Thanks!
- Applied many, many, many fixes from Markus Gothe of Sweden -- Thanks!


lft 3.2 / WhoB 3.2
----------------------
- Added support for 802.1q tagged VLANs
- Manual page corrections (thanks to Brett)
- WhoB will use Prefix WhoIs for bulk file resolution in
  Cymru-compatible format with '-c' but Cymru will be used 
  with '-C'


lft 3.1 / WhoB 3.1
----------------------
- New configure options:
  --enable-gtod Forces LFT to use gettimeofday() on each packet instead of 
  using the BPF timestamp. This is critical on platforms that have enabled
  'fastts' or that do not have high-precision BPF timestamping.
  --enable-universal generates binaries including both PPC and Intel architecture 
  (for users running Mac OS X/Darwin)
- Improved compatibility with NetBSD and Darwin/Mac OS X
- Added autoconf support for NetBSD
- Improved compatibility with older compilers (thanks to Sean)
- Updated autoconf bits and pieces
- By popular request, reversed the -g option of WhoB
  WhoB now uses gigo mode by default unless -g is specified which turns
  ON its parser and enables the other various options


lft 3.0
----------------------
- Completely refactored and now a workable library
- LFT has a new (-b) TCP Basic trace method that makes TCP traces NAT-friendly
- LFT has a new (-p) ICMP trace method that uses echo requests to trace paths
- Many memory issues fixed 
- Cleanup of several 2.6b5-3.0b features


lft 2.6 / WhoB 2.0
----------------------
- LFT prints 'open' in the target block if the target dest port is open
- LFT indicates the reason it marked ports open/closed in verbose(2) output 
- LFT prints an asterisk when it retransmits a packet after a timeout
- LFT uses Prefix WhoIs (bulk) or RISWHOIS (bulk) for netname resolution
- LFT has a new (-u) traditional UDP-based tracing feature
- LFT displays start and finish times/dates when using the (-T) option
- LFT has a new (-U) feature to display all times in UTC/GMT0
- LFT only shows time spent tracing/resolving at verbosity level 1 or higher
- LFT sets the ToS bit on outgoing IP datagrams when (-I) option is used
- LFT gets timevals from packets instead of calling gettimeofday()
- LFT won't be fooled into thinking there's a firewall on a gateway just
  because adaptive mode ups the state waiting for replies that never come
- Improved LFT performance (removed gettimeofday() on each packet)
- WhoB/library uses Prefix WhoIs to resolve OrgNames and NetNames
- WhoB/library has improved support for RIPE NCC RIS and Prefix WhoIs
- WhoB supports bulk resolution (-f option) from an input file
- WhoB also supports one-per-line output (-cf option) from a bulk input file
- WhoB will use putenv() on Solaris who is missing setenv() to set TZ
- WhoB has a new (-g) feature to take input directly from the command line 
  and print output directly from Prefix WhoIs (referred to as GIGO)
- Added $DESTDIR support to Makefile (thanks Daniel)
- Fixed an off-by-one bug in LFT related to ASN display encountered when 
  a trace contains one or more neglected TTLs
- Fixed an off-by-one bug in LFT's verbose output regarding TTLs
- Fixed variable ordering in whois.c from 2.6b1 (thanks Erick) 
- Numerous platform-specific improvements
- Reconfigured autoconf and segregated ./config/
- Updated autoconf components to v2.59


lft 2.5 / WhoB 1.5
----------------------
- Inclusive of betas 2.32 to 2.4x
- Added -z option to pseudo-randomize source port
- Added behavior to automatically select the most appropriate interface
  based on routing (this was on the most wanted list)
- Improved OpenBSD compatibility (IP length nonzero)
- OpenBSD is now detected by autoconf (for configuring the above)
- Darwin is now detected by autoconf and its definition disables
  some BSD features to make it compatible with MacOS X and Darwin
- LFT now indicates it has reached the target by printing a 'T'
  character in the status display (if status is enabled)
- Cleanups were made to the verbose output levels (-VVV)
- Significantly revamped whois framework makes it easy to include
  whois functionality into other programs
- Added -C and -R and -r options to force alternate ASN sources:
    - (r)IPE RIS - special thanks to Rene Wilhelm @ RIPE NCC
    - (C)ymru 
    - (R)ADB 
- Default ASN source (-A) is now Prefix WhoIs (see pwhois.org)
- LFT now queries for ASNs in bulk format after completing
  a trace if pwhois (default), RIPE NCC RIS, or Cymru is selected
- Added dst/src port autoselection based on user-supplied hostname
- Vastly improved standalone whois client "whob" see whob.8 (whob manpage)
- Makefile now installs 'whob' no-frills whois client (try ./whob)
- "Smart" mode is now referred to as "Adaptive" mode (-E)


lft 2.31
--------------------
- Fixed time precision on FreeBSD 5.3 (Thanks to Kurt Jaeger)


lft 2.3:
--------------------
- added WSAIoctl() call to select proper IF on windows based on dest (thanks Graham!)
- lowered max_retries to a default of 2
- cleaned up formatting related to -S option
- cleaned up verbose output to be more friendly

lft_types.h
/*
 *  lft_types.h
 *  Layer Four Traceroute
 *
 *  This file is part of LFT.
 *
 *  The LFT software provided in this Distribution is
 *  Copyright 2007 VOSTROM Holdings, Inc.
 *
 *  The full text of our legal notices is contained in the file called
 *  COPYING, included with this Distribution.
 *
 */
#ifndef LFT_TYPES_H
#define LFT_TYPES_H

#if defined( __CYGWIN__ ) || defined( WIN32 ) || defined(_WIN32)
#include "config/acconfig.win.h"
#else
#include "config/acconfig.h"
#endif
#include <assert.h>
#ifndef __FAVOR_BSD
# define __FAVOR_BSD	1
#endif
#if defined( __CYGWIN__ ) || defined( WIN32 ) || defined(_WIN32)
#define __USE_W32_SOCKETS
#include <Winsock2.h>
#include <windows.h>
#include <Ws2tcpip.h>
#include <sys/types.h>
#define LITTLE_ENDIAN 1
#define BYTE_ORDER 1
typedef signed long n_long;
typedef signed short n_short;
typedef signed long n_time;
#include <stdio.h>
#include <ctype.h>
#include <math.h>
#include <time.h>
#ifdef __CYGWIN__
# include <getopt.h>
#else
# include "include/win32/wingetopt.h"
# if defined(WIN32) || defined(_WIN32)
#  define SIZEOF_CHAR     1
#  define SIZEOF_SHORT    2
#  define SIZEOF_LONG     4
#  define SIZEOF_LONG_LONG    8
#  include "include/libpcap/bittypes.h"
#  include "include/libpcap/Gnuc.h"
#  define bzero(buf,cnt)    memset(buf,'\0',cnt);
# endif
#endif
#include "include/net/if_arp.h"
#include "include/netinet/if_ether.h"
#include "include/netinet/ip.h"
#include "include/netinet/ip_icmp.h"
#include "include/netinet/tcp.h"
#include "include/netinet/udp.h"
#else
#include <stdarg.h>
#include <sys/types.h>
#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

#include <stdio.h>
#include <ctype.h>
#include <math.h>
#include <sys/socket.h>
#include <net/if.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/if_ether.h>
#include <netdb.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <arpa/inet.h>
#include <errno.h>

#ifdef BSD
#include <machine/limits.h>
#endif

#ifdef BSD_IP_STACK
#include <sys/ioctl.h>
#include <net/bpf.h>
#if !defined(DARWIN) && !defined(NETBSD)
#define HAVE_SNPRINTF
#define HAVE_VSNPRINTF
#include <pcap-int.h>
#endif
#endif

#include <pcap.h>

#ifdef sun
#include <limits.h>
#include <strings.h>
#endif

#endif

#include "lft_ifname.h"
#include "lft_lsrr.h"
#include "whois.h"

#if defined(__FreeBSD__)
#include <sys/queue.h>
#elif !defined(DARWIN) && !defined(NETBSD)
#include "lft_queue.h"
#endif

#ifndef EXIT_FAILURE
#define EXIT_FAILURE 1
#endif

#ifndef EXIT_SUCCESS
#define EXIT_SUCCESS 0
#endif

#ifdef	__cplusplus
extern "C" {
#endif
/* holds the pseudo-header for generating checksums */
struct sumh
{
    unsigned int src;
    unsigned int dst;
    unsigned char fill;
    unsigned char protocol;
    unsigned short len;
};

/* The actual packet data */
struct trace_packet_s
{
    struct ip ip_hdr;
    struct ip_lsrr lsrr;		/* must go here for ip checksum to work */
    struct tcphdr tcp_hdr;
    struct udphdr udp_hdr;
    int size;
    char *payload;
    int payload_len;
};

/* RFC 1393 type trace IP option */
struct rfc1393_ip_option_s
{
	u_char optcode;		//=82
	u_char optlength;	//=12
	u_short id;			//number to identify icmp trace messages
	u_short ohc;			//outbound hop count
	u_short rhc;			//return hop count
	struct in_addr origip;		//originator ip address
} __attribute__((packed));
/* ICMP echo header */
struct icmp_echo_header_s
{
	u_char type;
	u_char code;
	u_short checksum;
	u_short id;
	u_short sequence;
} __attribute__((packed));
/* ICMP trace response header */
struct icmp_trace_reply_header_s
{
	u_char type;
	u_char code;
	u_short checksum;
	u_short id;
	u_short unused;
	u_short ohc;			//outbound hop count
	u_short rhc;			//return hop count
	u_long ols;				//output link speed
	u_long olmtu;			//output link MTU
} __attribute__((packed));
/* Trace packet for RFC 1393 type and ICMP base trace */
struct icmp_trace_packet_s
{
	char * packet;
	int packet_len;
	struct ip * ip_hdr;
	struct rfc1393_ip_option_s * icmp_trace_opt;
    struct ip_lsrr * lsrr;
	struct icmp_echo_header_s * echo;
    char * payload;
    int payload_len;
};

/* Packet container with additional info */
struct trace_packet_info_s
{
    int icmp_type;		/* ICMP_UNREACH code, -1 if RST reply */
    int is_done;		/* is this a final hop? */
    short hopno;
    unsigned int   seq;
    struct timeval sent;
    struct timeval recv;		/* 0 if unreceived */
    struct in_addr hopaddr;     /* IP address of router */
	union
	{
	    struct trace_packet_s packet;
		struct icmp_trace_packet_s icmp_packet;
	} u;
    SLIST_ENTRY(trace_packet_info_s) next_by_hop;
    SLIST_ENTRY(trace_packet_info_s) next;
};

/* hop information, by ttl */
struct hop_info_s
{
    int num_sent;
    int all_sent, all_rcvd;
    struct timeval ts_last_sent;
    struct timeval ts_last_recv;	
	struct trace_packet_info_s * done_packet;
    unsigned short state;
    unsigned short flags;
    SLIST_HEAD(hop_packets_s, trace_packet_info_s) packets;
};

#ifdef	__cplusplus
}
#endif

#endif /*LFT_TYPES_H*/


lft_lib.h
/*
 *  lft_lib.h
 *  Layer Four Traceroute
 *
 *  This file is part of LFT.
 *
 *  The LFT software provided in this Distribution is
 *  Copyright 2007 VOSTROM Holdings, Inc.
 *
 *  The full text of our legal notices is contained in the file called
 *  COPYING, included with this Distribution.
 *
 */
#ifndef LFT_LIB_H
#define LFT_LIB_H

#include "lft_types.h"
struct in_addr ipaddress;
/* not available in earlier darwin systems */
#ifndef AI_NUMERICSERV 
#define AI_NUMERICSERV 0 
#endif 

/* As the trace progresses, each hope will attempt
to work through the states one by one until it
receives an answer (2 attempts per state).
Whatever state "works" - will be then set up on 
following hops to continue from.
*/
#define HS_SEND_FIN         0x00
#define HS_SEND_SYN         0x01
#define HS_SEND_SYN_FIN     0x02    
#define HS_SEND_RST         0x04
#define HS_SEND_SYN_ACK     0x12
#define HS_SEND_ACK         0x16
#define HS_MAX              (HS_SEND_SYN)

#define HF_ENDPOINT         0x01

/* default timeout value */
#define DEFAULT_TIMEOUT_MS  250 

/* Common EtherType values */
#ifndef ETHERTYPE_IP
#define	ETHERTYPE_IP		0x0800	/* IP protocol */
#endif
#ifndef ETHERTYPE_ARP
#define	ETHERTYPE_ARP		0x0806	/* Addr. resolution protocol */
#endif
#ifndef ETHERTYPE_REVARP
#define	ETHERTYPE_REVARP	0x8035	/* reverse Addr. resolution protocol */
#endif
#ifndef ETHERTYPE_VLAN
#define	ETHERTYPE_VLAN		0x8100	/* IEEE 802.1Q VLAN tagging */
#endif
#ifndef ETHERTYPE_IPV6
#define	ETHERTYPE_IPV6		0x86dd	/* IPv6 */   
#endif

/* Sometimes-missing BPF values */
#ifndef DLT_RAW
#define DLT_RAW         101     /* Raw IP */
#endif
#ifndef DLT_PPP_SERIAL
#define DLT_PPP_SERIAL  50      /* PPP with HDLC encapsulation */
#endif
#ifndef DLT_PPP_ETHER
#define DLT_PPP_ETHER   51      /* PPP over Ethernet */
#endif
#ifndef DLT_LINUX_SLL
#define DLT_LINUX_SLL   113     /* Linux cooked capture */
#endif
#ifndef DLT_PPP
#define DLT_PPP         9       /* PPP over Ethernet */
#endif


/* ToS (type of service) bits we can set on the IP datagram */

#define TOSMINDELAY         0x10
#define TOSMAXTHROUGH       0x08
#define TOSMAXRELIABLE      0x04
#define TOSMINCOST          0x02

/*Errors and warnings codes*/
#define WRN_CANT_SETUP_FIN              -1
#define WRN_CANT_DISP_HOST_NAMES        -2
#define WRN_ADAPTIVE_DISABLED_BY_UDP    -3
#define WRN_FIN_DISABLED_BY_UDP         -4
#define WRN_ONLY_ONE_ASN_LOOKUP         -5
#define WRN_UDP_PORT_TOO_HIGH           -6
#define WRN_PACKET_LENGTH_TOO_HIGH      -7
#define WRN_PACKET_LENGTH_TOO_LOW       -8
#define WRN_CANT_DISABLE_RESOLVER       -9
#define WRN_ALREADY_RANDOM_SPORT        -10
#define WRN_ADAPTIVE_DISABLED_BY_FIN    -12
#define ERR_DEVNAME_TOO_LONG            -13
#define WRN_UNABLE_SETUP_UTC            -14

#define WRN_GETIFFORREMOTE_SOCKET       -15
#define WRN_GETIFFORREMOTE_CONNECT      -16
#define WRN_GETIFFORREMOTE_SOCKNAME     -17
#define ERR_UNKNOWN_HOST                -18
#define ERR_RAW_SOCKET                  -19
#define ERR_SOCKET_BIND                 -20
#define WRN_WSAIOCTL                    -21
#define ERR_IP_HDRINCL                  -22
#define ERR_NOT_ENOUGH_MEM              -23
#define ERR_RAW_TCP_DISABLED            -24

typedef struct _badhopstateparam
{
    const struct hop_info_s *h;
    short nhop;
}WrnBadHopStateParam;
#define WRN_BAD_HOP_STATE               -25
#define WRN_NS_LOOKUP_FAILED            -26
#define ERR_WIN_SELECT                  -27
#define ERR_WIN_RECV                    -28
#define ERR_WIN_WSASTARTUP              -29
#define ERR_PCAP_ERROR                  -30
#define ERR_DISCOVER_INTERFACE          -31
#define ERR_UNKNOWN_INTERFACE           -32
#define ERR_UNKNOWN_SEND_INTERFACE      -32
#define ERR_PCAP_DEV_UNAVAILABLE        -33
#define WRN_BIOCIMMEDIATE               -34
#define ERR_PCAP_NONBLOCK_ERROR         -35
/*Events codes and their params structures*/
#define EVT_AUTOCONFIGURED_TO_PORTS     1
#define EVT_ADDRESS_INITIALIZED         2
typedef struct _sentpacketparams
{
    short nhop;
    unsigned int tseq;
    unsigned char flags;
    unsigned short tttl;
}EvtSentPacketParam;
#define EVT_SENT_PACKET                 3
#define EVT_SHOW_PAYLOAD                4
#define EVT_SHOW_UDP_CHECKSUM           5
#define EVT_SHOW_TCP_CHECKSUM           6
#define EVT_SHOW_HOPS                   7
#define EVT_SHOW_NUM_HOPS               8
#define EVT_TRACE_COMPLETED             9
#define EVT_ON_RESOLUTION               10
#define EVT_TRACE_REPORT_START          11
typedef struct _rptnoreplyparams
{
    int hopno;
    int noreply;
}EvtNoReplyParam;
#define EVT_RPT_NO_REPLY                12
#define EVT_RPT_FRW_INSPECT_PACKS       13
#define EVT_RPT_FRW_STATE_FILTER        14
#define EVT_RPT_BSD_BUG                 15
#define EVT_RPT_HOP_INFO_START          16
typedef struct _packetinfoevtparam
{
    int asnumber;
    const char * netname;
    struct in_addr last_hop;
    const struct trace_packet_info_s * tp;
}EvtPacketInfoParam;
#define EVT_RPT_PACKET_INFO             17
#define EVT_RPT_PACKET_LIST_END         18
#define EVT_RPT_NO_HOPS                 19
#define EVT_RPT_TIME_TRACE              20
#define EVT_ON_EXIT                     21
#define EVT_TTL_NO_REPLY                22
#define EVT_PROGRESS_NO_REPLY           23
#define EVT_TTL_TOUT_RESEND             24
#define EVT_TTL_TOUT_GIVINGUP           25
typedef struct _debugchkpoint1
{
    int last_return;
    int no_reply;
    int need_reply;
}EvtDebugCheckpoint1Param;
#define EVT_DBG_CHECKPOINT1             26
#define EVT_CANT_RELIABLY_RTRIP         27
#define EVT_HAVE_UNANSWERRED_HOPS       28
#define EVT_TOO_FAR_AHEAD               29
#define EVT_HAVE_GAPS                   30
#define EVT_EITHER_RESP_OR_TOUT         31
#define EVT_LOOKFOR_UNINC_ACK           32
#define EVT_LOOKFOR_OFF_BY_LEN          33
#define EVT_LOOKFOR_LAST_RESORT         34
#define EVT_SKIP_PACKET                 35
typedef struct _nonseqpack
{
    struct in_addr ipaddr;
    const struct trace_packet_info_s * tp;
}EvtNonSeqPacketParam;
#define EVT_ACK_WAS_NOT_INC             36
#define EVT_RST_REL_TO_ISN              37
#define EVT_ACK_WAS_WAY_OFF             38
#define EVT_DUPLICATE_PACKET            39
#define EVT_PROGRESS_DUPLICATE          40
typedef struct _recvpacket
{
    struct in_addr ipaddr;
    const struct trace_packet_info_s * tp;
    unsigned int seq;
}EvtRecvPacketParam;
#define EVT_RECV_PACKET                 41
#define EVT_PROGRESS_OK                 42
#define EVT_TCP_PORT_CLOSED             43
#define EVT_TCP_PORT_OPEN               44
#define EVT_PROCESS_PACKET_START        45
#define EVT_UDP_NOT_FOR_US              46
typedef struct _incomudpicmp
{
    const struct ip * ip;
    const struct ip * orig_ip;
    const struct udphdr *udp;
    const struct icmp *icmp;
}EvtIncomingICMPUDPParam;
#define EVT_INCOMING_ICMP_UDP           47
#define EVT_RCVD_ICMP_UDP               48
typedef struct _incomtcpicmp
{
    const struct ip * ip;
    const struct ip * orig_ip;
    const struct tcphdr *tcp;
    const struct icmp *icmp;
}EvtIncomingICMPTCPParam;
#define EVT_INCOMING_ICMP_TCP           49
#define EVT_RCVD_ICMP_TCP               50
#define EVT_RCVD_TCP                    51
#define EVT_RCVD_UNKNOWN                52
#define EVT_DEVICE_SELECTED             53
#define EVT_SHOW_INITIAL_SEQNUM         54
#define EVT_TRACE_START                 55
#define EVT_DBG_CHECKPOINT2             56

#define EVT_DBG_LOG_MESSAGE             57

#define EVT_PROGRESS_SKIP_PACKET	58

#define EVT_OPEN_CHECK_RESULT		59

#define ERR_BTCP_PROBE_PORT_IS_BUSY	60
#define ERR_BTCP_WRONG_PORT_VALUE	61

#define EVT_OCHECK_START		62
#define WRN_OCHECK_OPEN_SOCK		63
#define WRN_OCHECK_IOCTL		64
#define WRN_OCHECK_SELECT		65
#define WRN_OCHECK_GETERROR		66
#define WRN_OCHECK_SOCKERROR		67
#define WRN_OCHECK_TIMEOUT		68
#define EVT_OCHECK_OPEN			69
#define WRN_OCHECK_FCNTLGET		70
#define WRN_OCHECK_FCNTLSET		71
#define WRN_OCHECK_CONNECTERR		72

typedef struct _incomechoreplyicmp
{
    const struct ip * ip;
    const struct icmp_echo_header_s * echo;
}EvtIncomingICMPEchoParam;
#define EVT_INCOMING_ICMP_Echo		73
#define EVT_RCVD_ICMP_Echo		74

typedef struct _incomicmpicmp
{
    const struct ip * ip;
	const struct icmp * icmp;
	const struct ip * orig_ip;
    const struct icmp_echo_header_s * echo;
}EvtIncomingICMPICMPParam;
#define EVT_INCOMING_ICMP_ICMP		75
#define EVT_RCVD_ICMP_ICMP		76

#if defined(BSD_IP_STACK) && !defined(OPENBSD)
#define SCREWED_IP_LEN
#endif

typedef struct btcpmapentry
{
	int nhop;
	int port;
	int sentcount;
}BasicTCPMapEntry;
#ifdef	__cplusplus
extern "C" {
#endif
typedef struct _btcp_debug_info
{
	int type;
	int hop;
	int phop;
	int port;
	struct in_addr ip;
}btcp_debug_info;
/* Session parameters */
typedef struct _lft_session_params
{
    struct timeval ts_last_sent;
    struct timeval now;
    double scatter_ms;                      /* milleseconds between sends */
    int ttl_min;                            /* user may request to start at a higher TTL */
    int hop_info_length;

    unsigned short ip_id;                   /*not used*/
    unsigned char tcp_flags;

    int use_fins;

    int seq_start;       /* generate ISN internally by default */
    int dport;           /* set default destination to tcp/80 HTTP */
    int sport;           /* set default source to tcp/53 dns-xfer */
    int auto_ports;      /* enable port autoselection by default */
    int random_source;   /* disable random source port by default */
    int set_tos;         /* disable set ToS bit by default */
    int userlen;         /* user-requested packet length */
    int payloadlen;      /* the final probe payloadlength */
    int win_len;

    int timeout_ms;      /* timeout between retries */
    int retry_max;       /* number of retries before giving up */
    int retry_min;       /* minimum number of checks per hop */
    int ahead_limit;     /* number of probes we can send
                          * without replies if we don't know
                          * the number of hops */
    int dflag;

    int ttl_limit;       /* max # hops to traverse (highest TTL) */
    int break_on_icmp;	 /* break on icmp other than time exceeded */
    int noisy;           /* disable verbose debug by default */
    int nostatus;        /* print status bar by default */
    int userdevsel;      /* by default, we'll select the device */
    int senddevsel;      /* by default, we'll select the device */
    int resolve_names;   /* dns resolution enabled by default */
    int hostnames_only;	 /* disable printing of IP addresses */
    int timetrace;       /* disable tracer timing by default */
    int adaptive;		 /* disable state engine by default */
	int protocol;		 /* 0 - TCP, 1 - UDP, 2 - ICMP base, 3 - ICMP RFC 1393, 4 - TCP basic */
    int do_netlookup;    /* disable netname lookup by default */
    int do_aslookup;     /* disable asn lookup by default */
    int use_radb;        /* use RADB instead of pwhois */
    int use_cymru;       /* use Cymru instead of pwhois */
    int use_ris;         /* use RIPE NCC RIS instead of pwhois */

    char *payload;

    int send_sock;
    int skip_header_len;

#if defined( __CYGWIN__ ) || defined( WIN32 ) || defined(_WIN32)
    int recv_sock;
    int wsastarted;
#else
    pcap_t * pcapdescr;
#endif
    int UseLocalTime;

    int num_hops;
    /*int num_sent;*/
    int num_rcvd;
    int target_open;
	int target_filtered;
    int target_anomaly;

    char *hostname;
    char *hostname_lsrr[9];
    int hostname_lsrr_size;

    struct in_addr local_address;
    struct in_addr remote_address;

    struct timeval begin_time, trace_done_time;

    /* The actual packet data (one of..)*/
    struct trace_packet_s trace_packet;
	struct icmp_trace_packet_s icmp_packet;

    /* Packet container with additional info */
    /* struct trace_packet_info_s * trace_packet_info;*/		/* indexed by dport - dport NOT USED*/

    /* list of packet containers */
    SLIST_HEAD(packets_s, trace_packet_info_s) trace_packets;
    int trace_packets_num;

	/* Map of ports for basic TCP trace */
	BasicTCPMapEntry * btcpmap;
	int latestmapchoice;
	int btcpmapsize;
	int btcpdpucnt;
	int trg_probe_is_sent;
	/* btcp_debug_info debugmap[1000]; */
	/* int debugmapidx; */
	
    /* hop information, by ttl */
    struct hop_info_s * hop_info;
    const char * pcap_dev;
    /* data link type as in pcap_datalink() */
    int pcap_datalink;
    const char * pcap_send_dev;
    const char * userdev;
    const char * senddev;
    /*WHOIS parameters*/
    whois_session_params * wsess;
    /*User's data*/
    void * UsersDataCookie;

    /*Exit status. When this field has value <0 lft will end work as soon as possible*/
    int exit_state;
}lft_session_params;

extern const char * icmp_messages[];
/*--------------------------- Callbacks definition ---------------------------*/
/*
Paramaters: 
    lft_session_params * sess - session handle, 
    int code - code of error or event,
    const void * param - additional parameters, depend on code
*/
typedef void (*LFT_CALLBACK)(lft_session_params *, int, const void *);
/*----------------------------------------------------------------------------*/
void LFTInitializeCallbacks(LFT_CALLBACK error_handler, LFT_CALLBACK event_handler);
lft_session_params * LFTSessionOpen(void);
void LFTSessionClose(lft_session_params * sess);
double timediff_ms (struct timeval prior, struct timeval latter);
unsigned int get_address(lft_session_params * sess, const char *host);
#ifndef SCREWED_IP_LEN 
u_int32_t ip_cksum (const struct ip *ip);
#endif
u_int32_t tcp_cksum (struct ip *ip, struct tcphdr *tcp, const char * payload, int payload_len);
int hop_state_up (lft_session_params * sess, short nhop);
int hop_state_copy(lft_session_params * sess, short nhop);
unsigned int new_seq(lft_session_params * sess);
/*----------------------------------------------------------------------------*/
/*                          Safe setting of parameters                        */
/*----------------------------------------------------------------------------*/
/*Use TCP FIN packets exclusively (defaults are SYN)*/
int LFTSetupFIN(lft_session_params * sess);
/*Display hosts symbolically; suppress IP address display*/
int LFTSetupDispSymbHost(lft_session_params * sess);
/*Use traditional UDP (probes) for tracing instead of TCP*/
int LFTSetupUDPMode(lft_session_params * sess);
#define ASN_LOOKUP_RIS      0
#define ASN_LOOKUP_RADB     1
#define ASN_LOOKUP_CYMRU    2
/*Use RIPE NCC's RIS to resolve ASNs instead of Prefix WhoIs*/
int LFTSetupRISLookup(lft_session_params * sess);
/*Use the RADB to resolve ASNs instead of Prefix WhoIs*/
int LFTSetupRADBLookup(lft_session_params * sess);
/*Use Cymru to resolve ASNs instead of Prefix WhoIs*/
int LFTSetupCYMRULookup(lft_session_params * sess);
/*Destination port number (same as using target:port as target)*/
int LFTSetupDestinationPort(lft_session_params * sess, char * userport);
/*Set the length of the probe packet in bytes*/
int LFTSetupLengthOfPacket(lft_session_params * sess, int plen);
/*Display hosts numerically; disable use of the DNS resolver*/
int LFTSetupDisableResolver(lft_session_params * sess);
/*Source port number*/
int LFTSetupSourcePort(lft_session_params * sess, int port);
/*Use LFT's stateful engine to detect firewalls and path anomalies*/
int LFTSetupAdaptiveMode(lft_session_params * sess);
/*Use a specific device by name or IP address (\"en1\" or \"1.2.3.4\")*/
int LFTSetupDevice(lft_session_params * sess,char * udev);
/*Use a specific device by name or IP address (\"en1\" or \"1.2.3.4\")*/
int LFTSetupSendDevice(lft_session_params * sess,char * sdev);
/*Display all times in UTC (GMT0).  Activates -T option automatically*/
int LFTSetupUTCTimes(lft_session_params * sess);
/*----------------------------------------------------------------------------*/
int lft_resolve_port (lft_session_params * sess, const char *strport);
void LFTExecute(lft_session_params * sess);
void lft_printf(lft_session_params * sess, const char *templ, ...);
/*----------------------------------------------------------------------------*/
void setOutputStyle(int nstyle); /* 0 - ordinary output, 1 - xml output, 2 - json*/
int outputStyleIsXML(void);
/*----------------------------------------------------------------------------*/
#ifdef	__cplusplus
}
#endif

#endif /*LFT_LIB_H*/
